### ISSUE TITLE
Type hint error when module and class shares the same name

### ISSUE THREAD

**Issue by fourdim** [microsoft/pyright#7233]
Type hint error when module and class shares the same name.

In `project/core/__init__.py`
```python
from .Binding import Binding
```

In `project/core/Binding.py`
```python
class Binding():
    def __init__(self) -> None:
        pass
```

In `project/core/test.py`
```python
import core
bindings: List[core.Binding] = []
# Error: Module cannot be used as a type
```

In `project/__init__.py`:
```python
from .core import *
```

But when running using python directly:
```python
>>> import core
>>> core.Binding
<class 'core.Binding.Binding'>
>>> core.Binding.Binding
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'Binding' has no attribute 'Binding'
```

Pylance: v2023.12.1

**Comment by erictraut** [comment]
I'm not able to repro the bug as reported. 

@fourdim, could you verify that if you follow these steps with the latest version of pylance that you still see the problem?

@debonte, @rchiodo, are you able to repro this problem in Pylance?

**Comment by fourdim** [comment]
> I'm not able to repro the bug as reported.

I'm so sorry that the import I've written is wrong.

The correct version will be:
In `project/core/test.py`
```python
from typing import List
from project import core
bindings: List[core.Binding] = []
# Error: Module cannot be used as a type
```

I have created a repro repo https://github.com/fourdim/pyright-7233

Please kindly check it.

**Comment by debonte** [comment]
@erictraut, I'm guessing that you're able to repro this now. It repros for me on Pyright 1.1.350.

The hover info is strange. Seems like hover sees `core.Binding` as the class, but the diagnostic sees it as the module.

![image](https://github.com/microsoft/pyright/assets/12820898/2a7484f2-f8e4-4cc2-981a-18e4ce2c2159)

**Comment by erictraut** [comment]
Yes, I'm able to repro now. One thing that wasn't clear from the original bug report is that "project" is not the project folder (i.e. the folder that is opened in VS Code). Instead, "project" is a subfolder within the project.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.352, which I just published.
